timelock-impl/src/main/java/com/palantir/atlasdb/timelock/LockTokenConvertingTimelockService.java

```diff
@@ -0,0 +1,147 @@
+        return delegate.getFreshTimestamp();
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(@QueryParam("number") int numTimestampsRequested) {
+        return delegate.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp(LockImmutableTimestampRequest request) {
+        LockImmutableTimestampResponse delegateResponse = delegate.lockImmutableTimestamp(request);
+        return ImmutableLockImmutableTimestampResponse.builder()
+                .from(delegateResponse)
+                .lock(makeSerializable(delegateResponse.getLock()))
+                .build();
+    }
+
+    @Override
+    public long getImmutableTimestamp() {
+        return delegate.getImmutableTimestamp();
+    }
+
+    @Override
+    public LockResponseV2 lock(LockRequestV2 request) {
+        LockResponseV2 delegateResponse = delegate.lock(request);
+        if (delegateResponse.wasSuccessful()) {
+            return ImmutableLockResponseV2.builder()
+                    .from(delegateResponse)
+                    .tokenOrEmpty(makeSerializable(delegateResponse.getToken()))
+                    .build();
+        }
+        return ImmutableLockResponseV2.copyOf(delegateResponse);
+    }
+
+    @Override
+    public WaitForLocksResponse waitForLocks(WaitForLocksRequest request) {
+        return delegate.waitForLocks(request);
+    }
+
+    @Override
+    public Set<LockTokenV2> refreshLockLeases(Set<LockTokenV2> tokens) {
+        return makeAllSerializable(delegate.refreshLockLeases(castAllToAdapters(tokens)));
+    }
+
+    @Override
+    public Set<LockTokenV2> unlock(Set<LockTokenV2> tokens) {
+        return makeAllSerializable(delegate.unlock(castAllToAdapters(tokens)));
+    }
+
+    @Override
+    public long currentTimeMillis() {
+        return delegate.currentTimeMillis();
+    }
+
+    private static Set<LockTokenV2> makeAllSerializable(Set<LockTokenV2> tokens) {
+        return tokens.stream()
+                .map(LockTokenConvertingTimelockService::makeSerializable)
+                .collect(Collectors.toSet());
+    }
+
+    @VisibleForTesting
+    static LockTokenV2 makeSerializable(LockTokenV2 token) {
+        return ImmutableLockTokenV2.copyOf(token);
+    }
+
+    private static Set<LockTokenV2> castAllToAdapters(Set<LockTokenV2> tokens) {
+        return tokens.stream()
+                .map(LockTokenConvertingTimelockService::castToAdapter)
+                .collect(Collectors.toSet());
+    }
+
+    @VisibleForTesting
+    static LockTokenV2 castToAdapter(LockTokenV2 token) {
+        LockRefreshToken legacyToken = new LockRefreshToken(
+                calculateLegacyTokenId(token.getRequestId()), Long.MIN_VALUE);
+        return new LegacyTimelockService.LockRefreshTokenV2Adapter(legacyToken);
+    }
+
+    private static BigInteger calculateLegacyTokenId(UUID uuid) {
+        return new BigInteger(ByteBuffer.allocate(16)
```

Comment:
(I've added a throw, but this is done in `LockRefreshTokenV2Adapter`'s token class - this one starts from the potentially-less-information state and builds the BigInteger from that.)